### PR TITLE
Specific exceptions

### LINKED ISSUE #167: Subclass RuntimeException in StandardSchedulerClient for better error handling
Due to `StandardSchedulerClient` throwing plain `RuntimeException` instances there is no way to distinguish (and possibly ignore) exceptions other than by doing string matching of the supplied message:
```
        public void cancel(TaskInstanceId taskInstanceId) {
...
            if (execution.isPresent()) {
                if (execution.get().isPicked()) {
                    throw new RuntimeException(String.format("Could not cancel schedule, the execution with name '%s' and id '%s' is currently executing", taskName, instanceId));
                }
...
            } else {
                throw new RuntimeException(String.format("Could not cancel schedule - no task with name '%s' and id '%s' was found.", taskName, instanceId));
            }
        }
```
Almost the same is done in `reschedule()`, except for slightly different message wording.

Did you consider having `TaskInstanceAlreadyRunningException` and `TaskInstanceNotFoundException` runtime exception classes? The messages can stay the same in case someone is already doing string matching for that purpose.

### PR DESCRIPTION
**Pull request by zendern**
Fixes #167 

Couple of things happened in this PR. 

1. Did the exceptions asked for in #167 (9548953)
2. Found other places we were doing RuntimeExceptions and switched those over to custom exceptions as well. (a720f7e & 507c78f)
3. Introduced mockito for unit testing as I was unsure how to easily make the exceptional cases occur and figured that was easiest path to get what we needed. 
4. Move the one custom exception (DataClassMismatchException) into new exceptions package (c33c2d6)

----

I did use an existing apache commons SerializationException which matches what we were doing but I'm open to creating our own if it feels odd that this library would emit an exception from apache commons. 

FWIW commons is coming in transitively from cronutils.
```
[INFO] --- maven-dependency-plugin:2.10:tree (default-cli) @ db-scheduler ---
[INFO] com.github.kagkarlsson:db-scheduler:jar:master-SNAPSHOT
[INFO] +- com.github.kagkarlsson:micro-jdbc:jar:0.3:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.7:compile
[INFO] +- com.cronutils:cron-utils:jar:9.1.3:compile
[INFO] |  +- org.glassfish:javax.el:jar:3.0.0:compile
[INFO] |  \- org.apache.commons:commons-lang3:jar:3.10:compile
```

### CODE REVIEW

**Comment by virtual-machinist** [comment]
Good job. :+1:  Couple of things I'd like to note -
1. `TaskInstanceCurrentlyRunningException` and `TaskInstanceNotFoundException` could have `taskName` and `instanceId` properties. Since message is already generated using them, it makes sense to expose them as properties for use cases where the original context is unavailable (eg. exception handlers) without message parsing. I'd also move message formatting inside the constructor that could accept those as parameters.
2. I'd use a library-specific `SerializationException` instead of using the one provided by a transitive (and shadowed) dependency.
3. Probably a question of taste, but I'd make an abstract parent class for all db-scheduler exceptions that directly extends `RuntimeException` with the needed constructors. This allows to differentiate db-scheduler exceptions in cases where a similar pattern is needed without enumerating all of them.

**Comment by kagkarlsson** [comment]
Thank you for contributing! Will have a look as soon as I can 👍

**Comment by zendern** [comment]
> Good job. 👍 Couple of things I'd like to note -
> 
> 1. `TaskInstanceCurrentlyRunningException` and `TaskInstanceNotFoundException` could have `taskName` and `instanceId` properties. Since message is already generated using them, it makes sense to expose them as properties for use cases where the original context is unavailable (eg. exception handlers) without message parsing. I'd also move message formatting inside the constructor that could accept those as parameters.
> 2. I'd use a library-specific `SerializationException` instead of using the one provided by a transitive (and shadowed) dependency.
> 3. Probably a question of taste, but I'd make an abstract parent class for all db-scheduler exceptions that directly extends `RuntimeException` with the needed constructors. This allows to differentiate db-scheduler exceptions in cases where a similar pattern is needed without enumerating all of them.

I believe all 3 of these should be address in 52d27e2. Please take a look and see if those changes look better.

**Comment by virtual-machinist** [comment]
@zendern Awesome. Thanks! Pretty much what I had in mind. :smiley_cat:

**Comment by kagkarlsson** [comment]
Started reviewing.

**Review by kagkarlsson — COMMENTED** [review]
Thanks for fixing this! Great having support for more specific exceptions.
Just some minor comments.

And thanks a lot for great suggestions @virtual-machinist

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskException.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+package com.github.kagkarlsson.scheduler.exceptions;
+
+public class TaskException extends DbSchedulerException {
+    static final long serialVersionUID = -2132850112553296790L;
+
+    private final String taskName;
+    private final String instanceId;
+
+    public TaskException(String message, String taskName, String instanceId, Exception ex){
```
[the remark this answers is not in the document]
Comment (zendern) [reply db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskException.java]: Good call. Cleaned up here d2a1d6c.

File: db-scheduler/src/test/java/com/github/kagkarlsson/scheduler/JdbcTaskRepositoryExceptionsTest.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    private String expectedTableName;
+
+    @BeforeEach
+    public void setup() throws NoSuchFieldException, IllegalAccessException {
+        expectedTableName = randomAlphanumeric(5);
+        jdbcTaskRepository = new JdbcTaskRepository(null, true, null, expectedTableName, null, null);
+
+        Field jdbcRunnerField = JdbcTaskRepository.class.getDeclaredField("jdbcRunner");
+        ReflectionUtils.makeAccessible(jdbcRunnerField);
+        jdbcRunnerField.set(jdbcTaskRepository, mockJdbcRunner);
```
[the remark this answers is not in the document]
Comment (zendern) [reply db-scheduler/src/test/java/com/github/kagkarlsson/scheduler/JdbcTaskRepositoryExceptionsTest.java]: done e624a6c

File: db-scheduler/src/test/java/com/github/kagkarlsson/scheduler/SchedulerClientExceptionsTest.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            Instant.now(),
+            new TaskInstance(taskInstance.getTaskName(), taskInstance.getId()),
+            true,
+            randomAlphanumeric(5),
+            null,
+            null,
+            0,
+            null,
+            1
+        );
```
[the remark this answers is not in the document]
Comment (zendern) [reply db-scheduler/src/test/java/com/github/kagkarlsson/scheduler/SchedulerClientExceptionsTest.java]: Helper method created here 8feff40

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskException.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.kagkarlsson.scheduler.exceptions;
+
+public class TaskException extends DbSchedulerException {
```
[the remark this answers is not in the document]
Comment (zendern) [reply db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskException.java]: Yah this makes sense and done here 18dea52

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskInstanceCurrentlyRunningException.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.kagkarlsson.scheduler.exceptions;
+
+public class TaskInstanceCurrentlyRunningException extends TaskException{
```
[the remark this answers is not in the document]
Comment (zendern) [reply db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskInstanceCurrentlyRunningException.java]: Yah no one like to 🏃‍♀️ anyways :) 1d9b255

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskInstanceCurrentlyRunningException.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    private static final long serialVersionUID = 6016933189040296444L;
+
+    public TaskInstanceCurrentlyRunningException(String taskName, String instanceId) {
+        super("Failed to perform action on task since it's currently running.", taskName, instanceId);
```
[the remark this answers is not in the document]
Comment (zendern) [reply db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskInstanceCurrentlyRunningException.java]: Implemented here 01fdf19

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskException.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    private final String instanceId;
+
+    public TaskException(String message, String taskName, String instanceId, Exception ex){
+        super(message, ex);
+        this.taskName = taskName;
+        this.instanceId = instanceId;
+    }
+
+    public TaskException(String message, String taskName, String instanceId){
+        super(message + String.format(" (task name: %s, instance id: %s)", taskName, instanceId));
```
[the remark this answers is not in the document]
Comment (zendern) [reply db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/exceptions/TaskException.java]: Fixed here 34f2896

**Comment by zendern** [comment]
@kagkarlsson I think that handles all comments. Thanks for the feedback. Let me know if there is anything else.

**Review by kagkarlsson — APPROVED** [review]
Great work 👍

**Comment by kagkarlsson** [comment]
Released in 10.3

**Pull request merged**